challenges-7-plus.yml

```diff
@@ -0,0 +1,277 @@
+          `turbo:before-cache` event doesn't execute properly.
+        c: |
+          This happenbed because we called `close` on our dialog and forgot to make the
+          `turbo:before-cache` callback asynchronous.
+        d: |
+          This happened because we called `close` on our dialog right inside
+          `turbo:before-cache` and the closing animation didn't have time to
+          finish before the cache was saved.
+      correct: d
+
+js_widget:
+  question: |
+    In this video, we've added a Weather Widget sanippet into our page and saw how it
+    only work the first time our page loaded. When we clicked on subsecuent pages, it
+    stopped working!
+    
+    Why is this so?
+  explanation: |
+    In our video, the snipped of code is executed every time we load a page because it's
+    in our `body` element! But in it there's a check before adding the `script` to our
+    `head`. This `script` is only added if it's not already present. This is why the
+    `script` tag was only added once, and therefore, only executed once between our page
+    views!
+    
+    > When we forced our snippet to add the `script` tag without any check, it effectively
+    > added a new copy of itself into our `head` *every* single time we change the page.
+    > This worked, but now we have a new problem: Our `head` is growing forever!
+  features:
+    multiple_choice:
+      choices:
+        a: |
+          Because the `script` tag added by our snippet of code is only added once, hence
+          it only executes once.
+        b: |
+          Because the snippet of code added by us is only executed once, hence it only
+          adds the `script` tag once
+        c: |
+          Because Turbo recognizes similar scripts between page views and avoids re-executing
+          them.
+        d: |
+          The widget snipped worked as expected. It did not stop working after subsecuent
+          page views.
+      correct: a
+
+analytics:
+  question: |
+    When does the event `turbo:before-render` fire?
+  explanation: |
+    This is very usefull because you can access the contents of the new `body` element
+    that is about to replace the old one, so that you can tweak it and modify it
+    before any JavaScript is executed!
+    
+    > You can access the "about to be added" `body` elements with `event.detail.newBody`!
+    
+    > Related is the `turbo:render` event, which fires *after* the entire rendering operation
+    > has completed.
+  features:
+    multiple_choice:
+      choices:
+        a: |
+          It fires every time turbo has to render a page, but before its new contents have
+          arrived via Ajax.
+        b: |
+          It fires once per session, before Turbo initializes its JavaScript engine.
+        c: |
+          It fires every time that turbo has received a new page via Ajax and before it
+          replaces the `body` element with the new `body` contents.
```

Comment:
I can't remember: does it fire on back/forward button? What about when a preview is temporarily shown? I might be interesting to mention these - even potentially make them part of the question/answer if it feels important.

Reply:
I rephrased it based on the information you provide on the video!